src/libostree/ostree-repo-checkout.c

```diff
@@ -436,10 +437,25 @@ checkout_file_hardlink (OstreeRepo                          *self,
             if (!glnx_fstatat (destination_dfd, destination_name, &dest_stbuf,
                                AT_SYMLINK_NOFOLLOW, error))
               return FALSE;
-            const gboolean is_identical =
+            gboolean is_identical =
               (src_stbuf.st_dev == dest_stbuf.st_dev &&
                src_stbuf.st_ino == dest_stbuf.st_ino);
+            if (!is_identical && (src_stbuf.st_size == dest_stbuf.st_size))
+              {
+                /* resort to a checksum check */
+                g_autofree char *abspath =
+                  glnx_fdrel_abspath (destination_dfd, destination_name);
+                g_autoptr(GFile) dest_file = g_file_new_for_path (abspath);
+
+                /* XXX: need an fd-relative version of this API */
+                g_autofree guchar *csum = NULL;
+                if (!ostree_checksum_file (dest_file, OSTREE_OBJECT_TYPE_FILE,
+                                           &csum, cancellable, error))
+                  return FALSE;
 
+                g_autofree char *actual_checksum = ostree_checksum_from_bytes (csum);
```

Comment:
(That said though there's no reason not to use `ostree_checksum_inplace_from_bytes()` and lose a malloc right?)